### ISSUE TITLE
[feature request] more documentation

### ISSUE THREAD

**Issue by turtleDev** [mourner/kdbush#14]
Hi !

Can you please document the tree-building process, as well as search and range queries?
especially the `select` process.

**Comment by mourner** [comment]
What kind of documentation would you want to see? More comments in the code? Or a section in the readme with a high-level description of the algorithms?

**Comment by turtleDev** [comment]
A little bit of both.
Things like how dataset is arranged (are data points only kept in node leaves vs data points in internal nodes?), how the tree is stored, and some general insights into design of your implementation.

**Comment by mourner** [comment]
I see, thanks! I'll try to add a bit of both. The main concept here is that the data is stored flat — it's a sorted array where median splits items into two halfs horizontally, then medians of both halfs split it into quarters vertically, etc. This makes it very fast, but also static.

**Comment by turtleDev** [comment]
do you think query performance would take a hit, if instead of using an array, you used an object based tree ?

**Comment by mourner** [comment]
Yes, both query performance and initial indexing would take much longer (judging by speedup compared to RBush, which is tree-based).

**Comment by hsnetzer** [comment]
I'd also appreciate a few comments in the code! I'm interested in porting to Swift...

**Comment by dfcreative** [comment]
@hsnetzer have a look at [kdgrass](https://github.com/dfcreative/kdgrass) - the fork

**Comment by mourner** [comment]
@dfcreative are there any substantial differences/improvements in your fork? The only difference I've noticed is slicing the input points, which is unnecessary since the array is not modified by the algorithm.

**Comment by dfcreative** [comment]
@mourner not at all, just docs and api rearrangements

**Comment by hsnetzer** [comment]
To answer the original question, the select function is basically Quicksort, right?

**Comment by mourner** [comment]
@hsnetzer it's related to quicksort, but does a different thing: https://en.wikipedia.org/wiki/Quickselect
